### PR TITLE
Counts stat sum

### PR DESCRIPTION
**Pull request by registerrier**
**Description**

This pull request adds a `sum` attribute to `CountStatistics` classes. It also adds an addition operator. This leaves the handling of e.g. `alpha` averaging to the class itself when the user wants to combine different measurements.
It also adds a `__getitem__` method to allow access to subarrays via masking.

This implementation could simplify a bit some statistics calculations in `MapDataset`.

**Dear reviewer**

### CODE REVIEW

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks a lot @registerrier, I have left a few inline comments to address, mostly on the add operator and indexing behavior.

In general I think we could actually introduce possible `CountsStatistic.info_dict()` as well, which returns the summary as a dict already. This could be useful for quick looks...

I was also wondering to maybe introduce a public API for `Dataset` -> `CountStatistics`. I don't have a strong opinion on this, but somehow I see `CountsStatistics` a bit similar in functionality of `Estimator` (just in terms of the quantities it computes), so maybe having something like `CountsStatistic.from_dataset()` would be a bit clearer, however `MapDataset.counts_statistic` could work as well, however it will e.g. not exist for `FluxPointsDataset` and it's definition might go away completely if we introduce a more flexible way fo defining fit statistics. In which case the tight coupling between `MapDataset` and `CountsStastistic` will go away.

**Comment by adonath** [comment]
@registerrier Do you have time to address the remaining comments?

File: gammapy/stats/counts_statistic.py [hunk 1]
```diff
@@ -208,6 +229,32 @@ def _n_sig_matching_significance_fcn(self, n_sig, significance, index):
...
+            raise TypeError(f"Cannot add CountsStatistic to {other.__class__}")
+
+        self.n_on += other.n_on
+        self.mu_bkg += other.n_bkg
+        return self
+
+    def __getitem__(self, key):
+        if self.n_on.shape != ():
+            return CashCountsStatistic(self.n_on[key], self.n_bkg[key])
+        elif key in [0, slice(None), slice(0,1)]:
```
[the remark this answers is not in the document]
Comment (registerrier) [reply gammapy/stats/counts_statistic.py]: Done.

File: gammapy/stats/counts_statistic.py [hunk 2]
```diff
@@ -158,6 +158,27 @@ def n_sig_matching_significance(self, significance):
             it.iternext()
         return n_sig
 
+    def sum(self):
```
[the remark this answers is not in the document]
Comment (registerrier) [reply gammapy/stats/counts_statistic.py]: Done

File: gammapy/stats/counts_statistic.py [hunk 3]
```diff
@@ -281,3 +328,36 @@ def _n_sig_matching_significance_fcn(self, n_sig, significance, index):
...
         )
         return np.sign(n_sig) * np.sqrt(np.clip(stat0 - stat1, 0, None)) - significance
+
+    def sum(self, axis=None):
+        n_on = self.n_on.sum(axis=axis)
+        n_off = self.n_off.sum(axis=axis)
+        alpha = self.n_bkg.sum(axis=axis)/n_off
+        return WStatCountsStatistic(n_on, n_off, alpha)
+
+    def __add__(self, other):
```
[the remark this answers is not in the document]
Comment (registerrier) [reply gammapy/stats/counts_statistic.py]: I finally removed this. The in-place `stack` method is problematic if this to be used internally with `Dataset` maps. One would avoid modification of the parent maps simply handling the `CountsStatistic`.

File: gammapy/stats/counts_statistic.py [hunk 4]
```diff
@@ -158,6 +158,24 @@ def n_sig_matching_significance(self, significance):
...
+             Axis or axes on which to perform the summation.
+             Default, axis=None, will perform the sum over the whole array.
+
+        Returns
+        -------
+        stat : `~gammapy.stats.CountsStatistics`
+             the return stat object
+        """
+
+    def stack(self, other):
```
Comment (adonath) [inline gammapy/stats/counts_statistic.py:176]: Remove...

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @registerrier  I have left one last minor inline comment.

**Review by adonath — APPROVED** [review]
Thanks @registerrier, I added one minor commit with some cleanup. Otherwise I have no further comments...

**Pull request merged**